src/ir/ir_traversing_visitor_with_result.h

```diff
@@ -0,0 +1,80 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+#ifndef SRC_IR_IR_TRAVERSING_VISITOR_WITH_RESULT_H_
+#define SRC_IR_IR_TRAVERSING_VISITOR_WITH_RESULT_H_
+
+#include <variant>
+#include "src/ir/ir_visitor.h"
+#include "src/ir/module.h"
+
+namespace raksha::ir {
+
+// A visitor that also traverses the children of a node and allows performing
+// different actions before (PreVisit) and after (PostVisit) the children are
+// visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
+template <typename Derived, typename Result=std::monostate>
+class IRTraversingVisitorWithResult : public IRVisitor<Derived, Result> {
+ public:
+  virtual ~IRTraversingVisitorWithResult() {}
+
+  // Invoked before all the children of `module` is visited.
+  virtual Result PreVisit(const Module& module) { return Result(); }
+  // Invoked after all the children of `module` is visited.
+  virtual void PostVisit(const Module& module, Result& result) {}
+  virtual void FoldResult(const Module& module, Result child_result, Result& result) {}
+  // Invoked before all the children of `block` is visited.
+  virtual Result PreVisit(const Block& block) { return Result(); }
+  // Invoked after all the children of `block` is visited.
+  virtual void PostVisit(const Block& block, Result& result) {}
+  virtual void FoldResult(const Block& block, Result child_result, Result& result) {}
+  // Invoked before all the children of `operation` is visited.
+  virtual Result PreVisit(const Operation& operation) { return Result(); }
+  // Invoked after all the children of `operation` is visited.
+  virtual void PostVisit(const Operation& operation, Result& result) {}
+  virtual void FoldResult(const Operation& operation, Result child_result, Result& result) {}
+
+  Result Visit(const Module& module) final override {
+    Result result = PreVisit(module);
+    for (const std::unique_ptr<Block>& block : module.blocks()) {
+      FoldResult(*block, block->Accept<Derived, Result>(*this), result);
+    }
+    PostVisit(module, result);
+    return result;
+  }
+
+  Result Visit(const Block& block) final override {
+    Result result = PreVisit(block);
+    for (const std::unique_ptr<Operation>& operation : block.operations()) {
+      FoldResult(*operation, operation->Accept<Derived, Result>(*this), result);
+    }
+    PostVisit(block, result);
+    return result;
+  }
+
+  Result Visit(const Operation& operation) final override {
+    Result result = PreVisit(operation);
+    const Module* module = operation.impl_module();
+    if (module != nullptr) {
+      FoldResult(operation, module->Accept<Derived, Result>(*this), result);
```

Comment:
Should this be `FoldResult(*module, module->Accept<Derived, Result>(*this), result)`?

Reply:
Similarly to above, `operation` is the parent, passed to ensure that, if different FoldResult functions are needed, they can be specialized based on the parent type. Passing the module in once via the visit and once via the fold doesn't seem useful to me.

Reply:
> Passing the module in once via the visit and once via the fold doesn't seem useful to me. 

It could be useful if you want to combine the result from a children in specific way depending upon what the child is? 

In any case, it was hard to grasp what `Fold` is happening. A simple fix might be to add a `parent_` prefix to argument name as in `parent_operation`.

Reply:
I also wonder if it makes sense to add a `Context` argument for the `Visit` methods.

Reply:
Is the context (in the class) not enough for this purpose already?

It would make it more functional and simplify testing (i.e. no need to inject values into the IRVisitor) but are there other benefits?

Reply:
I think we can change this later if we find a need.